tests/csv_datums/test_csv_sources.py

```diff
@@ -90,35 +106,157 @@ def source_interval():
+
+    def test_kraken_source_returns_validated_and_adapted_data(self, source, source_interval, requests, validation,
+                                                              make_json):
+        requests.set_get_response(json=make_json({'pair': [['data']]}))
+        adapted = AdaptedData(json=make_json({'pair': [['data']]}), with_interval=source_interval)
+        assert source.query(since=1559347200) == csv_datums_from(adapted)
+        assert validation.data == csv_datums_from(adapted)
+
+    def test_kraken_source_queries_data_up_to_last_incomplete_interval(self, source, source_interval, requests,
+                                                                       server_time, make_json):
+        server_time.set_current_time(1559347200 + source_interval * 3 * 60 + 10)
+        requests.set_get_responses(
+            make_json({'pair': [1]}, last=to_nano_sec(1559347200 + source_interval * 60)),
+            make_json({'pair': [2]}, last=to_nano_sec(1559347200 + source_interval * 1.5 * 60)),
+            make_json({'pair': [3]}, last=to_nano_sec(1559347200 + source_interval * 3 * 60)),
+            make_json({'pair': [4]}, last=to_nano_sec(1559347200 + source_interval * 3 * 60 + 1)),
+        )
+
+        adapted = AdaptedData(
+            json=make_json({'pair': [1, 2, 3]}, last=to_nano_sec(1559347200 + source_interval * 3 * 60)),
+            with_interval=source_interval)
+        assert source.query(since=1559347200) == csv_datums_from(adapted)
+
+    def test_subsequent_queries_are_paused_with_frequency(self, source, source_interval, requests, server_time,
+                                                          local_time, make_json):
+        server_time.set_current_time(1559347200 + source_interval * 2 * 60 + 10)
+        requests.set_get_responses(
+            make_json({'pair': [1]}, last=to_nano_sec(1559347200 + source_interval * 60)),
+            make_json({'pair': [2]}, last=to_nano_sec(1559347200 + source_interval * 2.5 * 60)),
+            make_json({'pair': [3]}, last=to_nano_sec(1559347200 + source_interval * 3 * 60)),
+        )
+
+        source.query(since=1559347200)
+        assert local_time.received_sleeps == [LEDGER_FREQUENCY, LEDGER_FREQUENCY]
+
+    def test_subsequent_queries_use_cache_instead_of_remote(self, source, source_interval, requests, server_time,
+                                                            local_time, make_json):
+        server_time.set_current_time(1559347200 + source_interval * 2 * 60 + 10)
+        requests.set_get_responses(
+            make_json({'pair': [[1, 0, 1559347201]]},
+                      last=to_nano_sec(1559347200 + source_interval * 60)),
+            make_json({'pair': [[2, 0, 1559347200 + source_interval + 10]]},
+                      last=to_nano_sec(1559347200 + source_interval * 2.5 * 60)),
+            make_json({'pair': [[3, 0, 1559347200 + source_interval * 2]]},
+                      last=to_nano_sec(1559347200 + source_interval * 3 * 60)),
```

Comment:
Integrate the timestamp setting and the last setting into the make_json or make_trades function